### PR TITLE
Tensorboard

### PR DESCRIPTION
**Pull request by hill-a**
Implemented Tensorboard capabilities into most models, and fixed a few issues found along the way.

**Tensorboard and bug fixes**

- added more documentation (some modules from common).
- added doc about using custom env
- added Tensorboard support for A2C, ACER, ACKTR, DDPG, DeepQ, PPO1, PPO2 and TRPO
- added episode reward to Tensorboard
- added documentation for Tensorboard usage
- added Identity for Box action space
- fixed render function ignoring parameters when using wrapped environments
- fixed PPO1 and TRPO done values for recurrent policies
- fixed image normalization not occurring when using images
- updated VecEnv objects for the new Gym version
- changed default policies for DDPG, so that DDPG now works correctly
- fixed DDPG behavior
- refactored DQN policies
- added registry for policies, can be passed as string to the agent
- added documentation for custom policies + policy registration
- fixed numpy warning when using DDPG Memory
- fixed DummyVecEnc observation buffer not being copied when stepping or reseting
- added ``deterministic`` argument in the predict function

### CODE REVIEW

File: stable_baselines/a2c/a2c.py [hunk 1]
```diff
@@ -143,8 +179,22 @@ def _train_step(self, obs, states, rewards, masks, actions, values):
             td_map[self.train_model.states_ph] = states
             td_map[self.train_model.masks_ph] = masks
 
-        policy_loss, value_loss, policy_entropy, _ = self.sess.run(
-            [self.pg_loss, self.vf_loss, self.entropy, self.apply_backprop], td_map)
+        if update % 10 == 9:
```
Comment (araffin) [inline stable_baselines/a2c/a2c.py]: Why 10? You should define it as a variable
```python
(update + 1) % 10 = 0
```

+ missing comments of what you are doing

File: stable_baselines/a2c/a2c.py [hunk 2]
```diff
@@ -155,14 +205,21 @@ def learn(self, total_timesteps, callback=None, seed=None, log_interval=100):
                                                     schedule=self.lr_schedule)
 
             runner = A2CRunner(self.env, self, n_steps=self.n_steps, gamma=self.gamma)
+            self.episode_reward = np.zeros((self.n_envs,))
 
             t_start = time.time()
             for update in range(1, total_timesteps // self.n_batch + 1):
-                obs, states, rewards, masks, actions, values = runner.run()
-                _, value_loss, policy_entropy = self._train_step(obs, states, rewards, masks, actions, values)
+                obs, states, rewards, masks, actions, values, true_reward = runner.run()
```
Comment (araffin) [inline stable_baselines/a2c/a2c.py]: Missing comment, what is true reward? I think this name is already use in gail

File: stable_baselines/a2c/utils.py [hunk 3]
```diff
@@ -557,3 +557,34 @@ def q_explained_variance(q_pred, q_true):
     _, var_pred = tf.nn.moments(q_true - q_pred, axes=[0, 1])
     check_shape([var_y, var_pred], [[]] * 2)
     return 1.0 - (var_pred / var_y)
+
+
+def total_episode_reward_logger(rew_acc, rewards, masks, writer, steps):
+    """
+    calculates the new running reward, and prints to tensorflow log the output
```
Comment (araffin) [inline stable_baselines/a2c/utils.py]: what is a running reward ?

File: stable_baselines/acktr/acktr_disc.py [hunk 4]
```diff
@@ -170,10 +206,22 @@ def _train_step(self, obs, states, rewards, masks, actions, values):
             td_map[self.train_model.states_ph] = states
             td_map[self.train_model.masks_ph] = masks
 
-        policy_loss, value_loss, policy_entropy, _ = self.sess.run(
-            [self.pg_loss, self.vf_loss, self.entropy, self.train_op],
-            td_map
-        )
+        if update % 10 == 9:
```
Comment (araffin) [inline stable_baselines/acktr/acktr_disc.py]: same remark as before

File: stable_baselines/ddpg/ddpg.py [hunk 5]
```diff
@@ -525,12 +563,30 @@ def _train_step(self):
...
+        td_map = {
             self.obs_train: batch['obs0'],
             self.actions: batch['actions'],
+            self.rewards: batch['rewards'],
             self.critic_target: target_q,
-        })
+            self.param_noise_stddev: 0 if self.param_noise is None else self.param_noise.current_stddev
+        }
+
+        if step % 100 == 0 and step >= 1:
```
Comment (araffin) [inline stable_baselines/ddpg/ddpg.py]: same remark

**Comment by hill-a** [comment]
- [ ] Update doc images

File: stable_baselines/common/vec_env/base_vec_env.py [hunk 6]
```diff
@@ -137,8 +137,11 @@ def step_wait(self):
     def close(self):
         return self.venv.close()
 
-    def render(self, mode='human'):
-        self.venv.render()
+    def render(self, **kwargs):
+        return self.venv.render(**kwargs)
+
+    def get_images(self):
```
Comment (araffin) [inline stable_baselines/common/vec_env/base_vec_env.py]: get_images is not defined in DummyVecEnv

File: stable_baselines/ddpg/ddpg.py [hunk 7]
```diff
@@ -675,6 +729,12 @@ def learn(self, total_timesteps, callback=None, seed=None, log_interval=100):
                             assert max_action.shape == action.shape
                             # scale for execution in env (as far as DDPG is concerned, every action is in [-1, 1])
                             new_obs, reward, done, _ = self.env.step(max_action * action)
+
+                            if writer is not None:
+                                self.episode_reward = total_episode_reward_logger(self.episode_reward,
+                                                                                  reward.reshape((1, 1)),
```
Comment (araffin) [inline stable_baselines/ddpg/ddpg.py]: reshape((1, 1)), not (1, -1) ?

File: stable_baselines/deepq/build_graph.py [hunk 8]
```diff
@@ -136,47 +134,40 @@ def build_act(make_obs_ph, q_func, num_actions, scope="deepq", reuse=None):
                 should be passed to outer variable scope
         and returns a tensor of shape (batch_size, num_actions) with values of every action.
     :param num_actions: (int) number of actions.
-    :param scope: (str or VariableScope) optional scope for variable_scope.
-    :param reuse: (bool) whether or not the variables should be reused. To be able to reuse the scope must be given.
+    :param observations_ph: (TensorFlow Tensor) the observation placeholder
+    :param stochastic_ph: (TensorFlow Tensor) the stochastic placeholder
```
Comment (araffin) [inline stable_baselines/deepq/build_graph.py]: ?

File: stable_baselines/deepq/build_graph.py [hunk 9]
```diff
@@ -136,47 +134,40 @@ def build_act(make_obs_ph, q_func, num_actions, scope="deepq", reuse=None):
...
+    :param update_eps_ph: (TensorFlow Tensor) the update_eps placeholder
```
Comment (araffin) [inline stable_baselines/deepq/build_graph.py]: ?

File: stable_baselines/deepq/build_graph.py (hunk 8 above)
Comment (hill-a) [reply stable_baselines/deepq/build_graph.py]: it changes nothing to the behaviour, it is only so the input placeholder is within the scope "input" so the graph is cleaner

File: stable_baselines/deepq/build_graph.py (hunk 9 above)
Comment (hill-a) [reply stable_baselines/deepq/build_graph.py]: same as the last comment: it changes nothing to the behaviour, it is only so the input placeholder is within the scope "input" so the graph is cleaner

File: stable_baselines/common/vec_env/base_vec_env.py (hunk 6 above)
Comment (hill-a) [reply stable_baselines/common/vec_env/base_vec_env.py]: Done

File: stable_baselines/common/base_class.py [hunk 10]
```diff
@@ -275,3 +279,39 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
+        self.writer = None
+
+    def __enter__(self):
+        if self.tensorboard_log_path is not None:
+            save_path = os.path.join(self.tensorboard_log_path,
+                                     "{}_{}".format(self.tb_log_name, self._get_latest_run_id() + 1))
+            self.writer = tf.summary.FileWriter(save_path, graph=self.graph)
+        return self.writer
+
+    def _get_latest_run_id(self):
```
[the remark this answers is not in the document]
Comment (hill-a) [reply stable_baselines/common/base_class.py]: Done

**Comment by araffin** [comment]
@hill-a I tagged it as a Release Candidate, just waiting to do performance checks (some atari env + other env) before complete release.

**Comment by hill-a** [comment]
- [ ] add deterministic kwarg to the predict function
